docs/notebooks/deep_gaussian_processes.pct.py

```diff
@@ -0,0 +1,355 @@
+F_MINIMIZER = MICHALEWICZ_2_MINIMUM
+
+search_space = Box([0, 0], [pi, pi])
+
+fig = plot_function_plotly(
+    function,
+    search_space.lower,
+    search_space.upper,
+    grid_density=100
+)
+fig.update_layout(height=800, width=800)
+fig.show()
+
+# %% [markdown]
+# ## Sample the observer over the search space
+#
+# We set up the observer as usual, using Sobol sampling to sample the initial points.
+
+# %%
+import trieste
+
+observer = mk_observer(function)
+
+num_initial_points = 20
+num_acquisitions = 20
+initial_query_points = search_space.sample_sobol(num_initial_points)
+initial_data = observer(initial_query_points)
+
+# %% [markdown]
+# ## Model the objective function
+#
+# The Bayesian optimization procedure estimates the next best points to query by using a probabilistic model of the objective. We'll use a two layer deep Gaussian process (DGP), built using GPflux. We also compare to a (shallow) GP.
+#
+# We note that the DGP model requires us to specify the number of inducing points, as we don't have the true posterior. We also have to use a stochastic optimizer, such as Adam. Fortunately, GPflux allows us to use the Keras `fit` method, which makes optimizing a lot easier!
+#
+# Since DGPs can be hard to build, Trieste provides some basic architectures: here we use the `build_vanilla_deep_gp` method.
+
+# %%
+from trieste.models.gpflux import GPfluxModelConfig, build_vanilla_deep_gp
+from gpflow.utilities import set_trainable
+
+
+def build_dgp_model(data):
+    variance = tf.math.reduce_variance(data.observations)
+
+    dgp = build_vanilla_deep_gp(data.query_points, num_layers=2, num_inducing=100)
+    dgp.f_layers[-1].kernel.kernel.variance.assign(variance)
+    dgp.f_layers[-1].mean_function = gpflow.mean_functions.Constant()
+    dgp.likelihood_layer.likelihood.variance.assign(1e-5)
+    set_trainable(dgp.likelihood_layer.likelihood.variance, False)
+
+    epochs = 200
+    batch_size = 100
+
+    optimizer = tf.optimizers.Adam(0.01)
+    fit_args = {
+        "batch_size": batch_size,
+        "epochs": epochs,
+        "verbose": 0,
+    }
+
+    return GPfluxModelConfig(**{
+        "model": dgp,
+        "model_args": {
+            "fit_args": fit_args,
+        },
+        "optimizer": optimizer,
+    })
+
+
+dgp_model = build_dgp_model(initial_data)
+
+# %% [markdown]
+# ## Run the optimization loop
+#
+# We can now run the Bayesian optimization loop by defining a `BayesianOptimizer` and calling its `optimize` method.
+#
+# The optimizer uses an acquisition rule to choose where in the search space to try on each optimization step. We'll start by using Thompson sampling.
+#
+# We'll run the optimizer for twenty steps. Note: this may take a while!
+# %%
+from trieste.acquisition.rule import DiscreteThompsonSampling
+
+bo = trieste.bayesian_optimizer.BayesianOptimizer(observer, search_space)
+acquisition_rule = DiscreteThompsonSampling(1000, 1)
+
+dgp_result = bo.optimize(num_acquisitions, initial_data, dgp_model,
```

Comment:
add a comment about tracking state, and that in the future it should be resolved